### PR TITLE
fix(app-shell/components): keyboard navigation of user settings menu

### PR DESCRIPTION
**Pull request by jonnybel**
#### Summary

This PR fixes the keyboard navigation on the user settings menu, so that the menu only opens when its toggle button is pressed (with `Enter` or `Space`) and doesn't remain open when the user navigates out of it.

It can also be closed with the `Esc` key.

Kapture 2020-10-07 at 12 30 01

### CODE REVIEW

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 1]
```diff
@@ -130,6 +136,15 @@ const getUserSettingsMenuItemLinkStyles = () => css`
...
 
 const UserSettingsMenuBody = (props: MenuBodyProps) => {
+  // Focus on a menu item when it's opened through keyboard
+  const menuElementRef = React.useRef(null);
+  const setFocus = () => {
+    menuElementRef.current?.focus();
+  };
+  React.useEffect(() => {
+    setFocus();
+  });
```
Comment (emmenko) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:146]: You're missing the dependency array.

**Review by emmenko — COMMENTED** [review]
Thanks!

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 2]
```diff
@@ -74,14 +74,20 @@ const UserAvatar = (
 };
 UserAvatar.displayName = 'UserAvatar';
 
-const stateReducer: DownshiftProps<{}>['stateReducer'] = (_state, changes) => {
+function getIsFocusedElementInMenu() {
+  return Boolean(
+    document?.activeElement?.attributes['data-user-settings-menu']
```
Comment (tdeekens) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:79]: Doesn't the document always exist?

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 3]
```diff
@@ -74,14 +74,20 @@ const UserAvatar = (
...
+
+const stateReducer: DownshiftProps<{}>['stateReducer'] = (state, changes) => {
   switch (changes.type) {
     // So in case the user wants to navigate with the tab button
     // we need to make sure that the menu does not close
     case Downshift.stateChangeTypes.blurButton:
       return {
         ...changes,
-        isOpen: true,
+        isOpen: state.isOpen && getIsFocusedElementInMenu() ? true : false,
```
Comment (tdeekens) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:90]:
```suggestion
        isOpen: state.isOpen && getIsFocusedElementInMenu() ?? false,
```

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 4]
```diff
@@ -130,6 +136,15 @@ const getUserSettingsMenuItemLinkStyles = () => css`
...
 
 const UserSettingsMenuBody = (props: MenuBodyProps) => {
+  // Focus on a menu item when it's opened through keyboard
+  const menuElementRef = React.useRef(null);
+  const setFocus = () => {
+    menuElementRef.current?.focus();
+  };
+  React.useEffect(() => {
+    setFocus();
+  }, []);
```
Comment (tdeekens) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:146]:
```suggestion
  React.useEffect(() => {
    menuElementRef.current?.focus();
  }, []);
```

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 5]
```diff
@@ -197,6 +222,8 @@ const UserSettingsMenuBody = (props: MenuBodyProps) => {
           target="_blank"
           rel="noopener noreferrer"
           onClick={() => props.downshiftProps.toggleMenu()}
+          data-user-settings-menu
+          ref={applicationsAppBarMenu ? undefined : menuElementRef}
```
Comment (tdeekens) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:226]:
```suggestion
          ref={!applicationsAppBarMenu ? menuElementRef : undefined}
```

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx (hunk 2 above)
Comment (jonnybel) [reply packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:79]: Unless the dog eats it 😆 

Yeah I think that's right, no need to do the check.

Comment (tdeekens) [reply packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:79]:
```suggestion
    document.activeElement?.attributes['data-user-settings-menu']
```

Here so you can commit them from GH :)

Comment (jonnybel) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:79]:
```suggestion
    document.activeElement?.attributes['data-user-settings-menu']
```

Comment (jonnybel) [reply packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx]: Already did the exact same thing 😄 🏎️

**Review by tdeekens — APPROVED** [review]

**Comment by jonnybel** [comment]
Am I missing something or is the CI having issues?

**Comment by emmenko** [comment]
Did you check the failing build?

image

**Comment by jonnybel** [comment]
Oh, sorry, I only saw the error on uploading the Cypress artifacts.
Thanks!

**Review by emmenko — APPROVED** [review]

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx [hunk 6]
```diff
@@ -137,7 +137,7 @@ const getUserSettingsMenuItemLinkStyles = () => css`
 
 const UserSettingsMenuBody = (props: MenuBodyProps) => {
   // Focus on a menu item when it's opened through keyboard
-  const menuElementRef = React.useRef<HTMLElement>(null);
+  const menuElementRef = React.useRef<HTMLAnchorElement>(null);
```
Comment (tdeekens) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.tsx:140]: ⚔️ TypeScript 🥊

**Pull request merged**